pkg/backend/filestate/state.go

```diff
@@ -185,7 +186,28 @@ func (b *localBackend) saveStack(name tokens.QName, snap *deploy.Snapshot, sm se
 
 	// And now write out the new snapshot file, overwriting that location.
 	if err = b.bucket.WriteAll(context.TODO(), file, byts, nil); err != nil {
-		return "", errors.Wrap(err, "An IO error occurred while writing the new snapshot file")
+
+		b.mutex.Lock()
+		defer b.mutex.Unlock()
+
+		// Retry the write 10 times in case of upstream bucket errors
+		_, _, err = retry.Until(context.TODO(), retry.Acceptor{
```

Comment:
I think you'll need to tune the `Delay`, `Backoff`, and `MaxDelay` options on this to match the GCS expectations. Maybe something like 1s, backoff of 1.2, max 30s?

Reply:
Yep, my intention is to tune this quite significantly. I wanted to get a feature branch out so that affected users could test the patch first, then we'll get ready for release from there. Does that sound like a reasonable approach?

Reply:
I've added this as suggested!